File: 강의 자료/02-알고리즘/12-트리/문제별 코드/1991-트리 순회/solution.java
import java.io.*;
import java.lang.reflect.Array;
import java.util.*;

public class Main {
    static FastReader scan = new FastReader();
    static StringBuilder sb = new StringBuilder();

    static int n;
    static int[][] childs;

    static void input() {
        n = scan.nextInt();
        childs = new int[30][2];
        for (int i = 0; i < n; i++) {
            char curCh = scan.next().charAt(0);
            int cur = (int)(curCh - 'A');
            for (int k = 0; k < 2; k++) {
                char ch = scan.next().charAt(0);
                if (ch != '.') childs[cur][k] = (int)(ch - 'A');
                else childs[cur][k] = -1;
            }
        }
    }

    static void in_order(int x) {
        if (x == -1) return;
        in_order(childs[x][0]);
        sb.append((char)(x + 'A'));
        in_order(childs[x][1]);
    }

    static void pre_order(int x) {
        if (x == -1) return;
        sb.append((char)(x + 'A'));
        pre_order(childs[x][0]);
        pre_order(childs[x][1]);
    }

    static void post_order(int x) {
        if (x == -1) return;
        post_order(childs[x][0]);
        post_order(childs[x][1]);
        sb.append((char)(x + 'A'));
    }

    static void pro() {
        pre_order(0);
        sb.append('\n');
        in_order(0);
        sb.append('\n');
        post_order(0);

        // 정답 출력하기
        System.out.print(sb);
    }

    public static void main(String[] args) {
        input();
        pro();
    }


    static class FastReader {
        BufferedReader br;
        StringTokenizer st;

        public FastReader() {
            br = new BufferedReader(new InputStreamReader(System.in));
        }

        public FastReader(String s) throws FileNotFoundException {
            br = new BufferedReader(new FileReader(new File(s)));
        }

        String next() {
            while (st == null || !st.hasMoreElements()) {
                try {
                    st = new StringTokenizer(br.readLine());
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
            return st.nextToken();
        }

        int nextInt() {
            return Integer.parseInt(next());
        }

        long nextLong() {
            return Long.parseLong(next());
        }

        double nextDouble() {
            return Double.parseDouble(next());
        }

        String nextLine() {
            String str = "";
            try {
                str = br.readLine();
            } catch (IOException e) {
                e.printStackTrace();
            }
            return str;
        }
    }
}

File: 강의 자료/02-알고리즘/12-트리/문제별 코드/1991-트리 순회/solution.py
import sys
from collections import deque
si = sys.stdin.readline
n = int(si())
childs = [[-1, -1] for _ in range(n)]

def input():
    for _ in range(n):
        x, l, r = si().strip().split()
        if l != '.': childs[ord(x) - ord('A')][0] = ord(l) - ord('A')
        if r != '.': childs[ord(x) - ord('A')][1] = ord(r) - ord('A')


def in_order(x):
    if x == -1: return
    in_order(childs[x][0])
    print(chr(x + 65), end='')
    in_order(childs[x][1])


def pre_order(x):
    if x == -1: return
    print(chr(x + 65), end='')
    pre_order(childs[x][0])
    pre_order(childs[x][1])


def post_order(x):
    if x == -1: return
    post_order(childs[x][0])
    post_order(childs[x][1])
    print(chr(x + 65), end='')


def pro():
    pre_order(0)
    print()
    in_order(0)
    print()
    post_order(0)


if __name__ == "__main__":
    input()
    pro()